### PR TITLE
feat: add app.beforeClose to register close function

### PR DESCRIPTION
**Pull request by popomore**
##### Checklist

- [ ] `npm test` passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines

##### Description of change

the function will be called when app.close

Ref https://github.com/eggjs/egg/issues/305

### CODE REVIEW

File: lib/egg.js [hunk 1]
```diff
@@ -157,21 +162,41 @@ class EggCore extends KoaApplication {
...
+  }
+
+  /**
+   * register a function that will be called when app close
+   * @param {Function} fn - the function that can be generator function or async function
+   * @return {Boolean} status - whether registered or not
+   */
+  onClose(fn) {
+    if (!is.function(fn)) return false;
+    this[CLOSESET].add(fn);
```
Comment (atian25) [inline lib/egg.js]: Set 放 fn 用什么判断唯一？

Comment (popomore) [inline lib/egg.js]: === 吧

File: test/fixtures/app-onclose/app.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  let count = 0;
+  function onlyOnce() {
+    if (count === 0) {
+      app.onlyOnce = false;
+    } else {
+      app.onlyOnce = true;
+    }
+    count++;
+  }
+  app.onClose(onlyOnce);
```
Comment (popomore) [inline test/fixtures/app-onclose/app.js]: @atian25 加了个用例

File: lib/egg.js [hunk 3]
```diff
@@ -157,21 +162,41 @@ class EggCore extends KoaApplication {
...
-    return new Promise((resolve, reject) => {
-      try {
-        this.emit('close');
-        this.removeAllListeners();
-        resolve();
-      } catch (err) {
-        reject(err);
+    const set = this[CLOSESET];
+    return co(function* () {
+      this.emit('close');
```
Comment (fengmk2) [inline lib/egg.js]: 应该在before close fn运行完成之后才触发close事件吧

File: lib/egg.js [hunk 4]
```diff
@@ -157,21 +162,41 @@ class EggCore extends KoaApplication {
...
+    }.bind(this));
+  }
+
+  /**
+   * register a function that will be called when app close
+   * @param {Function} fn - the function that can be generator function or async function
+   * @return {Boolean} status - whether registered or not
+   */
+  beforeClose(fn) {
+    if (!is.function(fn)) return false;
```
Comment (fengmk2) [inline lib/egg.js]: 直接抛异常，或者assert断言

File: lib/egg.js [hunk 5]
```diff
@@ -157,21 +162,41 @@ class EggCore extends KoaApplication {
...
+   * return base controller
+   * @member {Controller}
+   */
   get Controller() {
     return this.BaseContextClass;
   }
 
+  /**
+   * return base service
+   * @member {Server}
```
Comment (fengmk2) [inline lib/egg.js]: Service

**Comment by popomore** [comment]
都改了

**Comment by fengmk2** [comment]
@popomore 你来发版本吧

**Comment by popomore** [comment]
+ egg-core@1.7.0

File: package.json [hunk 6]
```diff
@@ -7,6 +7,7 @@
     "autod": "autod",
     "lint": "eslint lib test *.js",
     "test": "npm run lint && egg-bin test",
```
Comment (atian25) [inline package.json]: 后面那个调用 test-local

**Pull request merged**